src/arithmetic/modular_set.cc

```diff
@@ -263,17 +263,26 @@ class ModularSetAnalyzer::Impl :
     }
   }
   /*!
-   * \brief Create interect of two sets.
+   * \brief Create intersection of two sets.
    * \param a The left operand.
    * \param b the right operand.
    */
-  static Entry Intersect(Entry a, Entry b) {
-    // simple rule for now: pick higher constraints.
-    // TODO(team-team): Use extended euclidean algorithm.
-    if (a.coeff == 0) return a;
-    if (b.coeff == 0) return b;
-    if (a.coeff >= b.coeff) return a;
-    return b;
+  static Entry Intersect(Entry x, Entry y) {
+    int64_t n, m;
+    int64_t a = x.coeff, b = x.base, c = y.coeff, d = y.base;
+    int64_t gcd = ExtendedEuclidean(a, c, &n, &m);
+    int64_t v = d - b;
+    if (v % gcd == 0) {
+      n = v / gcd * n;
+      m = v / gcd * (-m);
+
+      Entry ret;
+      ret.coeff = a / gcd * c;
+      ret.base = BaseSimplify(n * a + b, ret.coeff);
```

Comment:
Should we add this simplification to the constructor of Entry and make sure `coeff > 0` and `base in [0, coeff]`

Reply:
We should. The only case base is negative is when coeff equals 0, in which case it indicates the Singleton. We also need to make sure the algorithm works for such negative constant case